File: .sdk-repos/orchestration-cluster-api-python/examples/job.py
# Compilable usage examples for job and job worker operations.
# These examples are type-checked during build to guard against API regressions.
from __future__ import annotations

from camunda_orchestration_sdk import (
    CamundaAsyncClient,
    CamundaClient,
    ConnectedJobContext,
    JobActivationRequest,
    JobCompletionRequest,
    JobCompletionRequestVariables,
    JobError,
    JobFailRequest,
    JobFailure,
    JobKey,
    SyncJobContext,
    WorkerConfig,
)


# region ActivateJobs
async def activate_jobs_example() -> None:
    async with CamundaAsyncClient() as client:
        result = await client.activate_jobs(
            data=JobActivationRequest(
                type_="payment-processing",
                timeout=30000,
                max_jobs_to_activate=5,
            )
        )

        for job in result.jobs:
            print(f"Job {job.job_key}: {job.type_}")
# endregion ActivateJobs


# region CompleteJob
def complete_job_example(job_key: JobKey) -> None:
    client = CamundaClient()

    client.complete_job(
        job_key=job_key,
        data=JobCompletionRequest(
            variables=JobCompletionRequestVariables.from_dict(
                {"paymentId": "PAY-123", "status": "completed"}
            )
        ),
    )
# endregion CompleteJob


# region FailJob
def fail_job_example(job_key: JobKey) -> None:
    client = CamundaClient()

    client.fail_job(
        job_key=job_key,
        data=JobFailRequest(
            retries=2,
            error_message="Payment gateway timeout",
            retry_back_off=5000,
        ),
    )
# endregion FailJob


# region CreateJobWorker
async def create_job_worker_example() -> None:
    async with CamundaAsyncClient() as client:
        config = WorkerConfig(
            job_type="payment-processing",
            job_timeout_milliseconds=30_000,
            max_concurrent_jobs=5,
        )

        async def handler(job: ConnectedJobContext) -> dict[str, object]:
            job.log.info(f"Processing job {job.job_key}")
            return {"processed": True}

        client.create_job_worker(config=config, callback=handler)

        # Workers run continuously until cancelled
        await client.run_workers()
# endregion CreateJobWorker


# region JobWorkerWithErrorHandling
async def job_worker_with_error_handling_example() -> None:
    async with CamundaAsyncClient() as client:
        config = WorkerConfig(
            job_type="email-sending",
            job_timeout_milliseconds=60_000,
            max_concurrent_jobs=10,
        )

        async def handler(job: ConnectedJobContext) -> dict[str, object]:
            # Raise JobError to throw a BPMN error
            if not job.variables:
                raise JobError("MISSING_DATA", "No variables provided")

            # Raise JobFailure to fail with custom retries/backoff
            recipient = job.variables.to_dict().get("recipient")
            if not recipient:
                raise JobFailure(
                    "Missing recipient address",
                    retries=0,
                    retry_back_off=0,
                )

            print(f"Sending email for job {job.job_key}")
            return {"sent": True}

        client.create_job_worker(config=config, callback=handler)
        await client.run_workers()
# endregion JobWorkerWithErrorHandling


# region SyncJobWorkerCallback
async def sync_job_worker_callback_example() -> None:
    async with CamundaAsyncClient() as client:
        config = WorkerConfig(
            job_type="cpu-intensive-task",
            job_timeout_milliseconds=120_000,
        )

        # Sync callbacks run in a thread pool by default
        def handler(job: SyncJobContext) -> dict[str, object]:
            # CPU-bound or blocking I/O work
            return {"result": "computed"}

        client.create_job_worker(config=config, callback=handler)
        await client.run_workers()
# endregion SyncJobWorkerCallback

File: .sdk-repos/orchestration-cluster-api-js/examples/job.ts
// Compilable usage examples for job operations.
// These examples are type-checked during build to guard against API regressions.

import {
  createCamundaClient,
  type JobActionReceipt,
  type JobKey,
} from '@camunda8/orchestration-cluster-api';

//#region ActivateJobs
async function activateJobsExample() {
  const camunda = createCamundaClient();

  const result = await camunda.activateJobs({
    type: 'payment-processing',
    timeout: 30000,
    maxJobsToActivate: 5,
  });

  for (const job of result.jobs) {
    console.log(`Job ${job.jobKey}: ${job.type}`);

    // Each enriched job has helper methods
    await job.complete({ paymentId: 'PAY-123' });
  }
}
//#endregion ActivateJobs

//#region CompleteJob
async function completeJobExample(jobKey: JobKey) {
  const camunda = createCamundaClient();

  await camunda.completeJob({
    jobKey,
    variables: {
      paymentId: 'PAY-123',
      status: 'completed',
    },
  });
}
//#endregion CompleteJob

//#region FailJob
async function failJobExample(jobKey: JobKey) {
  const camunda = createCamundaClient();

  await camunda.failJob({
    jobKey,
    retries: 2,
    errorMessage: 'Payment gateway timeout',
    retryBackOff: 5000,
  });
}
//#endregion FailJob

//#region CreateJobWorker
async function createJobWorkerExample() {
  const camunda = createCamundaClient();

  const _worker = camunda.createJobWorker({
    jobType: 'payment-processing',
    jobTimeoutMs: 30000,
    maxParallelJobs: 5,
    jobHandler: async (job): Promise<JobActionReceipt> => {
      console.log(`Processing job ${job.jobKey}`);
      return job.complete({ processed: true });
    },
  });

  // Workers run continuously until closed
  // worker.close();
}
//#endregion CreateJobWorker

//#region JobWorker
async function jobWorkerExample() {
  const camunda = createCamundaClient();

  const _worker = camunda.createJobWorker({
    jobType: 'payment-processing',
    jobTimeoutMs: 30000,
    maxParallelJobs: 5,
    jobHandler: async (job): Promise<JobActionReceipt> => {
      const orderId = job.variables.orderId;
      console.log(`Processing order: ${orderId}`);

      // Complete the job with result variables
      return job.complete({ paymentId: 'PAY-123' });
    },
  });

  // Workers run until closed
  // worker.close();
}
//#endregion JobWorker

//#region JobWorkerWithErrorHandling
async function jobWorkerWithErrorHandlingExample() {
  const camunda = createCamundaClient();

  const worker = camunda.createJobWorker({
    jobType: 'email-sending',
    jobTimeoutMs: 60000,
    maxParallelJobs: 10,
    pollIntervalMs: 300,
    jobHandler: async (job): Promise<JobActionReceipt> => {
      try {
        console.log(`Sending email for job ${job.jobKey}`);
        return job.complete({ sent: true });
      } catch (err) {
        return job.fail({
          errorMessage: String(err),
          retries: (job.retries ?? 1) - 1,
        });
      }
    },
  });

  void worker;
}
//#endregion JobWorkerWithErrorHandling

//#region CreateThreadedJobWorker
async function createThreadedJobWorkerExample() {
  const camunda = createCamundaClient();

  const _worker = camunda.createThreadedJobWorker({
    jobType: 'cpu-heavy-task',
    handlerModule: './my-handler.js',
    maxParallelJobs: 32,
    jobTimeoutMs: 30000,
  });
}
//#endregion CreateThreadedJobWorker

//#region GetWorkers
function getWorkersExample() {
  const camunda = createCamundaClient();

  camunda.createJobWorker({
    jobType: 'payment-processing',
    jobHandler: async (job): Promise<JobActionReceipt> => job.complete(),
  });

  const workers = camunda.getWorkers();
  console.log(`Active workers: ${workers.length}`);
}
//#endregion GetWorkers

//#region StopAllWorkers
function stopAllWorkersExample() {
  const camunda = createCamundaClient();

  camunda.createJobWorker({
    jobType: 'payment-processing',
    jobHandler: async (job): Promise<JobActionReceipt> => job.complete(),
  });

  // Stop all workers and terminate thread pool
  camunda.stopAllWorkers();
}
//#endregion StopAllWorkers

//#region GetBackpressureState
function getBackpressureStateExample() {
  const camunda = createCamundaClient();

  const state = camunda.getBackpressureState();
  console.log(`Severity: ${state.severity}, Permits: ${state.permitsMax}`);
}
//#endregion GetBackpressureState

// Suppress "declared but never read"
void activateJobsExample;
void completeJobExample;
void failJobExample;
void createJobWorkerExample;
void jobWorkerExample;
void jobWorkerWithErrorHandlingExample;
void createThreadedJobWorkerExample;
void getWorkersExample;
void stopAllWorkersExample;
void getBackpressureStateExample;
